src/modm/math/proportional_unsigned.hpp

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2021, Thomas Sommer
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#pragma once
+
+#include <algorithm>
+
+#include <modm/math/utils/integer_traits.hpp>
+#include <modm/architecture/interface/assert.hpp>
+
+namespace modm {
+
+/**
+ * @brief 		Unsigned integer with arbitrary digits and scaling value on conversion
+ * 				between instances with different digits.
+ *
+ * @tparam D 	Number of Digits
+ *
+ * @author		Thomas Sommer
+ * @ingroup		modm_math
+ */
+template<int D>
+requires (D > 0)
+class ProportionalUnsigned {
+public:
+	static constexpr int Digits = D;
+
+	using T = uint_t<D>::least;
+	static constexpr T min = 0;
+	static constexpr T max = bitmask<D>();
+
+	constexpr ProportionalUnsigned() = default;
+
+	constexpr ProportionalUnsigned(T value) {
+		if(std::is_constant_evaluated())
+			this->value = std::min(value, max);
+		else {
+			modm_assert_continue_fail_debug(value <= max, "ProportionalUnsigned", "value out of range");
+			// Constructing runtime should a.f.a.p.
+			this->value = value;
+		}
```

Comment:
I think clamping by default without error is fine here. I'm uncomfortable with only sometimes getting a runtime error.

```cpp
constexpr ProportionalUnsigned(T value) value(std::min(value, max)) {}
```

Reply:
In the meanwhile, I think std::clamp is bad and there really should be a switch to enable/disable the assert:

As long as `ProportionalUnsigned` used the right way, Ts range shouldn't be exceeded and if happens, its an issue! An "out of range" notification to the developer is much better than silently clamping the passed value.

In normal program flow this assert should - by any means -  be omitted: The constructor of ProportionalUnsigned is called for every single pixel on Buffer conversions!